File: javascript-testing/collision-avoidance/rayTracing.js
function rayTraceCollision(/*const int[]*/aircraft,/*const int[]*/obstacle,/*int*/radius,/*int[]*/collisionPoint){
    /**
     * This function will transfer over to C++
     * Here the program will detect if there is a possible
     * collision assuming a linear trajectory. TRUE means there is a collision.
     * Parameters: 
     * - aircraft is an object with the aircraft's current GPS coordinates and bounding radius, in the properties:
     * lat, lon, alt, radius, prevLat, prevLon, prevAlt
     * - obstacle is an object with the obstacle's current GPS coordinates and bounding radius, in the properties:
     * lat, lon, alt, radius
     * If a collision is detected, the array collisionPoint
     * will be updated with the coordinates of the projected collision.
     */
    
    // aircraft - [ 0 => lon, 1 => lat, 2 => alt, 3 => prevLon, 4 => prevLat, 5 => prevAlt ]
    
    // This creates a ray that goes beyond the obstacle 
    var dlon = aircraft[0] - aircraft[3];
    var dlat = aircraft[1] - aircraft[4];
    var dalt = aircraft[2] - aircraft[5];

    cx.beginPath();
    cx.moveTo(aircraft[0],aircraft[1]);
    cx.lineTo(1500,aircraft[1]);
    cx.strokeStyle="#f00";
    cx.stroke();
    cx.closePath();
    
    // Calculate whether a collision occurs. Ray tracing calculation is provided by:
    // www.ccs.neu.edu/home/fell/CSU540/programs/RayTracingFormulas.htm
    
    var a = dlon*dlon + dlat*dlat + dalt*dalt;
    var b = 2*dlon*(aircraft[3]-obstacle[0]) + 2*dlat*(aircraft[4]-obstacle[1]) + 2*dalt*(aircraft[5] - obstacle[2]);
    var c = Math.pow(aircraft[3]-obstacle[0],2) + Math.pow(aircraft[4]-obstacle[1],2) + Math.pow(aircraft[5]-obstacle[2],2)-Math.pow(radius,2);
    
    
    var discriminant = Math.pow(b,2) - (4*a*c);
    var collision = discriminant >= 0;
    if ( collision ){
	var t1 = (-b-Math.sqrt(Math.pow(b,2)-4*a*c))/(2*a);
	var t2 = (-b+Math.sqrt(Math.pow(b,2)-4*a*c))/(2*a);
	if ( t1 > 0 ){
	    var t = t1;
	}
	else {
	    if ( t2 > 0 ){
		var t = t2;
	    }
	}
	collisionPoint[0] = aircraft[3] + t*dlon;
	collisionPoint[1] = aircraft[4] + t*dlat;
	collisionPoint[2] = aircraft[5] + t*dalt;
	return true;
    }
    else {
	return false;
    }
}
    

File: ray_tracing.c
/**
 * This file implements the ray tracing collision detection function that is
 * used to detect collisions with obstacles that are stationary and moving
 * along the same trajectory line as our aircraft.
 */

// aircraft[] and obstacle[] will contain the same elements:
// current position, previous position, and radius. so:
// aircraft[6] = {x,y,z,xprev,yprev,zprev,radius};
bool ray_tracing(const int aircraft[], const int obstacle[], int collision_point[]){
  int t  = -1;
  int dx = aircraft[0] - aircraft[3];
  int dy = aircraft[1] - aircraft[4];
  int dz = aircraft[2] - aircraft[5];

  int a = dx*dx + dy*dy + dz*dz;
  int b = 2*dx*(aircraft[3]-obstacle[0])+2*dy*(aircraft[4]-obstacle[1])+2*dz*(aircraft[5]-obstacle[2]);
  int c = (aircraft[3]-obstacle[0])^2+(aircraft[4]-obstacle[1])^2+(aircraft[5]-obstacle[2])^2-(aircraft[6]+obstacle[6])^2;

  int discriminant = b^2 - 4*a*c;
  if ( discriminant >= 0 ){
    int t1 = (-b-sqrt(b^2-4*a*c))/(2*a);
    int t2 = (-b+sqrt(b^2-4*a*c))/(2*a);
    if ( t1 > 0 ){
      t = t1;
    }
    else if ( t2 > 0 ){
      t = t2;
    }
    collision_point[0] = aircraft[3] + t*dx;
    collision_point[1] = aircraft[4] + t*dy;
    collision_point[2] = aircraft[5] + t*dz;
    return true;
  }
  else {
    return false;
  }
}
